Rework Inductor support for collectives.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101023
* #98643
* #98642
* __->__ #99765

This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @bertmaher @soumith @anijain2305 @desertfire

kumpera committed:
Rework Inductor support for collectives.

This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

[ghstack-poisoned]

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wanchaol commented on torch/_inductor/ir.py:
```diff
@@ -4108,21 +4119,40 @@ def codegen(self, wrapper):
-    Extract partial output from a multi-output OP.
-        Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
-    """
+class InPlaceCollectiveKernel(CollectiveKernel):
```
can we add some documentation about what's a `InPlaceCollectiveKernel` and `OutPlaceCollectiveKernel`, and what's their difference? It would also be beneficial to explain when to extend which kernel so that new collectives adding to inductor can have some guidance which to extend.

kumpera replied:
Good point. I added documentation to those two

wanchaol commented on torch/_inductor/ir.py:
```diff
@@ -4170,71 +4199,28 @@ def create(
 
 
-class AllReduce(CollectiveKernel):
+class AllReduce(InPlaceCollectiveKernel):
```
does it mean that `AllReduce` would by default be inplace node in the inductor IR? what if there're two consumers to the same node, one is to `allreduce`, the other is to do some other operations with the tensor content before `allreduce`, would this become a problem?

```
tensor=...
sumed = allreduce(tensor)
prod = tensor.prod()
```

kumpera replied:
That's 99.5% of the complication of this design. The point is enabling reuse whenever it's safe.

For now, there are basically 2 reasons on why it's unsafe for all_reduce to reuse the inputs, first if it's a graph input and second if the input tensor is used again after all_reduce.

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented on torch/_inductor/ir.py:
```diff
@@ -4390,35 +4386,25 @@ def __init__(self, layout, inputs, constant_args, reduce_op):
-        return AllReduce(
-            layout=new_layout,
-            inputs=[x],
+        _ = AllReduce(
```
what motivates using InplaceHint as a wrapper over the inputs and then not returning the inplace AllReduce op? I'm seeing the change as being equivalent so far and trying to understand how it is different

wconstab commented on torch/_inductor/ir.py:
```diff
@@ -4447,15 +4433,8 @@ def codegen_collective(self, wrapper, output_name, input_names):
             f"{output_name}, {input_names[0]}, async_op=True, group={output_name}_pg)"
         )
 
-        # At this point, output_name points to a fresh buffer
```
is my diff  view screwed up? it looks like there was a duplicate call to dist.* under this codegen_collective and you're just cleaning it up.  (but also removing the register_work part)

wconstab replied:
ok, switched to just reading the ir.py - makes a lot more sense now. nice design.

wconstab reviewed (approved):
lgtm- design is much cleaner!

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera commented on torch/_inductor/ir.py:
```diff
@@ -4390,35 +4386,25 @@ def __init__(self, layout, inputs, constant_args, reduce_op):
     def codegen_collective(self, wrapper, output_name, input_names):
-        # We must copy our input buffer sometimes, but the scheduler will help us find opportunities
-        # to reuse the input buffer.  (This requires no other users of the input buffer.)
-        if not wrapper.did_reuse(self, self.inputs[0]):
```
Precisely. :)

kumpera committed:
Update on "Rework Inductor support for collectives."


This is done by introducing two new base classes: InPlaceCollectiveKernel and OutOfPlaceCollectiveKernel.

They deal with the differences for when InPlaceHint needs to be used.

Additionally to that, we introduce `has_side_effects` method to buffers that
prevents them from being DCE'd by the scheduduler. This is needed because InPlaceHint
nodes both wrap the inputs and are the outputs, which places no users to the collectives
themselves.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge

kumpera commented:
@pytorchmergebot merge -f "Windows failure unrelated"
